tests/e2e/kserve_test.go

```diff
@@ -416,3 +423,144 @@ func (tc *KserveTestCtx) validateTemplatedResourceOwnerRefsAndLabels(expectOwned
 			WithCustomErrorMsg(msg, child.gvk.Kind, child.nn.Name, child.nn.Namespace))
 	}
 }
+
+// ValidateDSCIDSCValidationInteractionForKserve tests DSCI and DSC validation interaction during reconciliation.
+func (tc *KserveTestCtx) ValidateDSCIDSCValidationInteractionForKserve(t *testing.T) {
+	t.Helper()
+
+	jqKserveReadyFalse := `
+	.status.conditions[]
+	| select(.type == "KserveReady" and .status == "%s")
+	| .message
+	| test("servicemesh needs to be set to 'managed' in dsci cr"; "i")`
+
+	originalDSCI := tc.FetchDSCInitialization()
+
+	t.Log("Disabling ServiceMesh in DSCI")
+	tc.EnsureResourceCreatedOrPatched(
+		WithMinimalObject(gvk.DSCInitialization, tc.DSCInitializationNamespacedName),
+		WithMutateFunc(testf.Transform(
+			`.spec.serviceMesh.managementState = "%s"`, operatorv1.Removed,
+		)),
+	)
+
+	t.Log("Verifying KServe reports dependency on ServiceMesh correctly")
+	tc.EnsureResourceExists(
+		WithMinimalObject(gvk.DataScienceCluster, tc.DataScienceClusterNamespacedName),
+		WithCondition(And(
+			jq.Match(jqKserveReadyFalse, metav1.ConditionFalse),
+		)),
+		WithCustomErrorMsg("KServe should report ServiceMesh dependency and kserve readiness correctly"),
+		WithEventuallyTimeout(tc.TestTimeouts.longEventuallyTimeout),
+	)
+
+	t.Log("Re-enabling ServiceMesh in DSCI for recovery")
+	tc.EnsureResourceCreatedOrPatched(
+		WithMinimalObject(gvk.DSCInitialization, tc.DSCInitializationNamespacedName),
+		WithMutateFunc(testf.Transform(
+			`.spec.serviceMesh.managementState = "%s"`,
+			originalDSCI.Spec.ServiceMesh.ManagementState,
+		)),
+	)
```

Comment:
If we’re re-enabling ServiceMesh, fetching the original DSCI may not be necessary, as it should already have ServiceMesh enabled at this stage. Additionally, if the goal is simply to ensure it’s enabled, reapplying the enablement is harmless and will still work even if ServiceMesh is already active.

Reply:
@MarianMacik  As per this discussion, if the test fails, the E2E test will be stopped. So @ugiordan doesn't want the defer function to keep the logic the same everywhere.

Reply:
@MarianMacik If a test fails at any stage, the entire E2E suite will report that failure. So, recovering resources doesn’t really make sense in this context, as the tests should stop and not continue with the rest of the suite.